[launch] Refactoring ExecuteProcess into Execute and Executable (#272)
ros2/design: articles/execute_process_refactor.md

> # Refactoring ExecuteProcess in ROS2

Reviewer: nit: space between title and paragraph

Author: Addressing in upcoming revision.

> Initially proposed in [114](https://github.com/ros2/launch/issues/114), this document describes changes to `ExecuteProcess` and related classes. As initially implemented, some aspects of this are tightly coupled to both the implementation chosen and the assumption that processes will execute only in the local environment of the launch process itself. The following changes are designed to provide a more decoupled architecture that will ease support of launching processes locally, remotely, or potentially in containerized environments. Additionally, these changes will support the goal of decoupling the description of an executable from the actual action of executing it. These changes are designed so that existing systems may continue to use the interface before these changes without modification; newer systems may find benefits from adapting to the newer syntax, particularly when composing more complicated environments.

Reviewer: nit: one sentence per line. It makes reviewing easier.

Author: Addressing in upcoming revision.

> Initially proposed in [114](https://github.com/ros2/launch/issues/114), this document describes changes to `ExecuteProcess` and related classes. As initially implemented, some aspects of this are tightly coupled to both the implementation chosen and the assumption that processes will execute only in the local environment of the launch process itself. The following changes are designed to provide a more decoupled architecture that will ease support of launching processes locally, remotely, or potentially in containerized environments. Additionally, these changes will support the goal of decoupling the description of an executable from the actual action of executing it. These changes are designed so that existing systems may continue to use the interface before these changes without modification; newer systems may find benefits from adapting to the newer syntax, particularly when composing more complicated environments.

Reviewer: nit: space between end of paragraph and next title.

Author: Addressing in upcoming revision.

| `launch.events.process.ProcessStarted` | emitted when the process starts                              |
| `launch.events.process.ProcessExited`  | emitted when the process exits; event contains return code   |
| `launch.events.process.ProcessStdout`  | emitted when the process produces data the stdout pipe; event contains the data from the pipe |
| `launch.events.process.ProcessStderr`  | emitted when the process produces data the stderr pipe; event contains the data from the pipe |

### launch_ros.actions.Node

Extends the `ExecuteProcess` class. This allows execution of a ROS-specific node, and extends the available information to include node related details.

#### Constructor

Added constructor parameters for this class all focus on concepts similar to the Launch Details discussed above, but the class also overrides the `execute` method of `ExecuteProcess`. The purpose of this override is to provide more granular structure for the command used to launch a ROS node. All the provided constructor parameters eventually reduce into a collection of substitution elements to compose the `cmd` parameter of `ExecuteProcess`.

|Argument|Description|
|---|---|

> |node_executable|the name of the executable to find if a package is provided or otherwise a path to the executable to run.|

Reviewer 2: @roger-strain nit: here and everywhere else, capitalize each description sentence.

Author: A lot of these were copied and pasted from the current class docs, so I was just maintaining how they appeared there. Easy enough change, however; addressing in upcoming revision.

| ----------------------------------------- | ------------------------------------------------------------ |
| `launch.events.lifecycle.StateTransition` | this event is emitted when a message is published to the "`/<node_name>/transition_event`" topic, indicating the lifecycle node represented by this action changed state |

## Considerations for Refactored Classes

The primary goal of the refactor is to attempt to split the definition aspect of processes/nodes apart from the logic which executes those elements. `ComposableNode` provides a basis for similar breakouts of other elements.

In general, existing classes should have similar analogs in the refactor to ensure that existing classes can be replaced with backwards-compatible replacements that use the new structure. These replacements can be implemented to warn the user that the older definitions are deprecated, and suggest moving to the newer refactored style.

Further consideration will be required for alternate frontends, such as XML, to determine how to implement these changes with streamlined syntax where possible.

> One difficulty when looking forward to allowing functionality such as remote and containerized execution is that the target environment is likely different than the local environment. Different packages may be installed, different environment variables may be defined, and so forth. To support this, it may be helpful to allow an alternate `launch.LaunchContext` to be defined for specific executions. Additionally, the available support for concepts such as signals and PIDs may be different; to that end, it is likely that remote execution of processes will require sibling classes to those described below, rather than direct subclasses.

Reviewer: I don't think an alternate version of launch context is needed.
I think we just need alternate "remote" versions of substitutions when it applies.
e.g.: `launch.substitutions.EnvironmentVariable` will need an alternate version.

There are two possibilities: add another substitutions for remotely executed processes, e.g.: `launch.remote.substitutions.EnvironmentVariable`. Or either add an extra argument to the existing substitution to specify to be configured in the remote machine.

Author: I'll rework this section a little bit to talk more toward the ability of a subclass implementing remote or containerized processes to override the method which handles substitutions. The exact method of doing it isn't supremely relevant at this moment. A full-fledged launch context may be unnecessary, but we need some way to address that problem, so just calling out where to do it is probably preferable.

## Proposed New Classes

> ### launch.descriptions.Process

Reviewer 2: @roger-strain I think we're mixing things up again here e.g. a process is yet another execution detail that may or may not apply to a node. The fact that some of its arguments may be dependent on the actual execution environment sets this one apart from the rest. `launch.descriptions.Executable` or `launch.descriptions.Program` maybe? We can later compose things to serve common use cases.

|env|dictionary of environment variables to be used, starting from a clean environment. If `None`, the current environment is used.|
|additional\_env|dictionary of environment variables to be added. If `env` was `None`, they are added to the current environment. If not, `env` is updated with `additional_env`.|

#### Properties

Accessors would be provided for the various constructor parameters: `cmd`, `name`, `cwd`, `env`, and `additional_env`.

#### Methods

In the current classes, substitutions are handled through internal method calls. To support better isolation of class functionality, substitutions will be performed in this class, but exposed from a public method which can be overridden by subclasses to perform additional appropriate substitutions.

| Name          | Description                                                  |
| ------------- | ------------------------------------------------------------ |
| apply_context | Takes a given LaunchContext and performs actions necessary to prepare to execute the defined command in that context. This will primarily consist of expanding various substitutions. |

#### Events

No events would be handled nor emitted.

> ### launch_ros.descriptions.Node

Reviewer 2: Hmm, there can be many `Node`s in a single process and even be dynamically composable. I'd imagine a `ROSAwareExecutable` that can bear `Node`s.

|package|name of the ROS package the node executable lives in|
|node_name| name the node should have                                    |
|node_namespace|namespace the node should create topics/services/etc in|
|parameters| list of either paths to yaml files or dictionaries of parameters |
|remappings|list of from/to pairs for remapping names|
|arguments|container specific arguments to be passed to the node|

#### Properties

Accessors would be provided for the various constructor parameters: `node_executable`, `package`, `node_name`, `node_namespace`, `parameters`, `remappings`, and `arguments`. Inherited accessors would also be available: `cmd`, `name`, `cwd`, `env`, and `additional_env`.

#### Methods

This class would override the `apply_context` method of `launch.descriptions.Process` to apply additional ROS-specific substitutions, then invoke the superclass method itself. This function would match that of the current `launch_ros.actions.Node` internal method `_perform_substitutions`.

#### Events

No events would be handled or emitted.

### launch_ros.descriptions.ComposableNode

> This class would be modified to inherit from `launch_ros.descriptions.Node`. 

Reviewer: I don't think this is possible, as for example `node_executable` argument doesn't apply for a composable node.
Maybe, both descriptions can inherit from a common base description.

Reviewer 2: Also, `launch_ros.descriptions.ComposableNode` is not a kind of  `launch_ros.descriptions.Process`.

Reviewer 2: Following above's comment logic, I'd imagine a `Node` (e.g. a `ContainerNode`) in a `ROSAwareExecutable` bearing `ComposableNode`s. Of course, then we can provide typical aggregations like `ROSAwareExecutable(ContainerNode(*composable_nodes))`.

#### Constructor

Most parameters would be passed to the new superclass.

|Argument|Description|
|---|---|
|node_plugin|name of the plugin to be loaded|

Additional parameters that may be passed, which are handled by `launch_ros.descriptions.Node`: `package`, `node_name`, `node_namespace`, `parameters`, `remappings`, `arguments`. Note that the parameter `node_executable` would not be passed to the superclass, as it is not applicable to this object.

#### Properties

Accessors would be provided for the additional constructor parameter: `node_plugin`. Inherited accessors would also be available: `cmd`, `name`, `cwd`, `env`, `additional_env`, `node_executable`, `package`, `node_name`, `node_namespace`, `parameters`, `remappings`, and `arguments`.

#### Methods

No custom methods would be defined or overridden.

#### Events

No events would be handled or emitted.

> ### launch.actions.ExecuteLocalProcess

Reviewer 2: @roger-strain again I think we're mixing things up. I'd imagine e.g. `Execute` and `SSHExecute` actions only that can take `Executable` descriptions, delegating all description-specific details to the descriptions themselves.

## Considerations for Refactored Classes

The primary goal of the refactor is to attempt to split the definition aspect of processes/nodes apart from the logic which executes those elements.
`ComposableNode` provides a basis for similar breakouts of other elements.

In general, existing classes should have similar analogs in the refactor to ensure that existing classes can be replaced with backwards-compatible replacements that use the new structure.
These replacements can be implemented to warn the user that the older definitions are deprecated, and suggest moving to the newer refactored style.

Further consideration will be required for alternate frontends, such as XML, to determine how to implement these changes with streamlined syntax where possible.

One difficulty when looking forward to allowing functionality such as remote and containerized execution is that the target environment is likely different than the local environment.
Different packages may be installed, different environment variables may be defined, and so forth.

> An subclass implementation of the proposed `launch.descriptions.Executable` class may choose to override the base `apply_context` method to implement additional layer(s) of substitutions appropriate to its destination contexts, such as remote installation paths or environment variables.

Reviewer: I'm still unsure if the concept of overriding `apply_context` is needed.
I don't see much benefit on it

Author: @ivanpauno The intent of allowing overrides of `apply_context` is so an executable in a more specific domain can perform domain-specific substitutions. Basically, at this level, we're in a context that has no knowledge of ROS or any ROS-specific concepts. We can do some level of substitution here, but can't take everything into consideration. To support those ROS-specific concepts, we have the `launch_ros.descriptions.RosExecutable` class. It would override `apply_context` so that it could pass substitution information down to the `launch_ros.descriptions.Node` objects, and possibly perform other actions necessary for preparing the ROS executable to run.

My fear is that if we don't do things this way, then we either have to make the `launch` namespace aware of ROS-specific concepts like nodes, or we alternately need to come up with some other more generic framework to ensure all aspects of the definition have an opportunity to do pre-launch setup (primarily substitutions). If there's another approach I'm missing that doesn't require the override, I'd be very happy to discuss it and look for a different take.

Reviewer: >  The intent of allowing overrides of apply_context is so an executable in a more specific domain can perform domain-specific substitutions

Can you give an example of a domain specific substitution?
I'm trying to understand the use case, but I'm failing to see how this will help.

> so that it could pass substitution information down to the launch_ros.descriptions.Node objects, and possibly perform other actions necessary for preparing the ROS executable to run.

Can you give an example of what "other actions" can be.

---

You can still go ahead with a proposed implementation with this comment unsolved, it's not a big deal.
But I want to understand how this detail is supposed to help.

Author: @ivanpauno It might be best to show you the current functionality I'm trying to find a place for. In the current setup, `launch_ros.actions.Node` is a subclass of `launch.actions.ExecuteProcess`. `ExecuteProcess` defines the `execute` method, which is overridden by `Node`. Part of what `Node` does with the override is applying the various substitutions by calling `_perform_substitutions(LaunchContext)`. Some of those substitutions are ROS-specific, some may not be. (Honestly, I'm going to have to be very careful in that area to make sure I split things out appropriately.) Those substitutions still need to be possible after the refactor.

The purpose of *this* call is to allow a subclass (in this case, `launch_ros.descriptions.RosExecutable`) to inject any additional substitution logic. For instance, the proposed `launch.descriptions.Executable` has absolutely no reason to know anything at all about a `launch_ros.descriptions.Node`'s Name property, but it still needs to have an opportunity to perform substitutions. I'll admit to not being 100% on all the functionality going on down there, but my feeling is that I need some path to make sure everything can be covered. This was my best effort at ensuring that path exists when it's needed.

Reviewer: Oh, I see what you mean now.

You need a way to "normalize" all `launch.descriptions.Executable` derived classes, in a way that `launch.actions.ExecuteLocal` or other actions can actually execute them.

The proposal makes sense to me!!

| cwd             | The directory in which to run the executable                 |
| env             | Dictionary of environment variables to be used, starting from a clean environment. If `None`, the current environment is used. |
| additional\_env | Dictionary of environment variables to be added. If `env` was `None`, they are added to the current environment. If not, `env` is updated with `additional_env`. |

#### Properties

Accessors would be provided for the various constructor parameters: `cmd`, `name`, `cwd`, `env`, and `additional_env`.

#### Methods

In the current classes, substitutions are handled through internal method calls.
To support better isolation of class functionality, substitutions will be performed in this class, but exposed from a public method which can be overridden by subclasses to perform additional appropriate substitutions.
Substituted versions of the various parameters will be stored in a local dictionary, similar to the current `__process_event_args`.

| Name          | Description                                                  |
| ------------- | ------------------------------------------------------------ |

> | apply_context | Takes a given LaunchContext and performs actions necessary to prepare to execute the defined command in that context. This will primarily consist of expanding various substitutions. |

Reviewer 2: @roger-strain please, could you remind me why we have `apply_context` _and_ `prepare` (as opposed to just `prepare`)? If it's for (semantic) consistency's sake, I can see several variations of `prepare` already.

Author: @hidmic I don't know if there's a specific reason for both. It might have evolved from building from both ends, if you will. If you'd rather simplify that down to just one or the other, I can definitely look at doing that.

Reviewer 2: The latter sounds good.

| prepare | Takes a given LaunchContext and Action, and performs actions necessary to prepare the node for execution in that context. This will primarily consist of expanding various substitutions, but may additionally add event handlers related to the Action which will be invoked. Additionally, this will call the `prepare()` method of any defined traits. |

This function would match that of the current `launch_ros.actions.Node` internal method `_perform_substitutions`.

#### Events

No events would be handled or emitted.

### launch_ros.descriptions.ComposableNode

This class would extend the `launch_ros.descriptions.Node` class, and provide the additional information required for defining a node which can be launched in a composable context. 

#### Constructor

Most parameters would be passed to the superclass.

|Argument|Description|
|---|---|
| package        | Optional. Name of the ROS package the node plugin lives in. If not specified, the package of the containing executable will be assumed. This may fail if context regarding the executable is not available.        |
|node_plugin|Name of the plugin to be loaded|

> Additional parameters that may be passed, which are handled by `launch_ros.descriptions.NodeBase`: `node_name`, `node_namespace`, `parameters`, `remappings`, `arguments`, `traits`.

Reviewer: I guess that `launch_ros.descriptions.NodeBase` is what was previously named `launch_ros.descriptions.Node`?

Author: Good catch, will update. Should be `launch_ros.descriptions.Node`; `NodeBase` was in a previous iteration, and has been removed.

| prepare | Overridden. Adds subscription event handlers to listen for ROS node lifecycle state changes; this handler will trigger appropriate events from this class when transitions are detected. |

#### Events

##### Handled Events

| Event Type                            | Description                                                  |
| ------------------------------------- | ------------------------------------------------------------ |
| `launch.events.lifecycle.ChangeState` | This event can be targeted to a single lifecycle node, or more than one, or even all lifecycle nodes, and it requests the targeted nodes to change state, see its documentation for more details |

##### Emitted Events

| Event Type                                | Description                                                  |
| ----------------------------------------- | ------------------------------------------------------------ |
| `launch.events.lifecycle.StateTransition` | This event is emitted when a message is published to the "`/<node_name>/transition_event`" topic, indicating the lifecycle node represented by this action changed state |

> ### launch_ros.traits.ComposesNodes

Reviewer: ComposesNodes sound weird to me.
It sounds like a ComposableNode container (?)

Author: Yes, that's the function here. But in trying to go with the "traits" concept, I was looking for naming what was more along the lines of "a thing this class does". I'm certainly open to changing the name if there's something that sounds better. I was a little worried that going straight with `launch_ros.traits.ComposableNodeContainer` might make it seem like this is the thing itself, rather than describing an aspect of the `launch_ros.descriptions.Node`.

Reviewer 2: I agree with @roger-strain. I can't think of a better name either. Ultimately, there will be a handy `ComposableNodeContainer` action that hides this away whenever using `rclcpp_components` executables, so I don't see a risk of confusion.

| Name    | Description                                                  |
| ------- | ------------------------------------------------------------ |
| prepare | Takes a given LaunchContext and Action, and performs actions necessary to prepare the node for execution in that context. This will primarily consist of expanding various substitutions, but may additionally add event handlers related to the Action which will be invoked. Additionally, this will call the `prepare()` method of any defined traits. |

This function would match that of the current `launch_ros.actions.Node` internal method `_perform_substitutions`.

#### Events

No events would be handled or emitted.

### launch_ros.descriptions.ComposableNode

This class would extend the `launch_ros.descriptions.Node` class, and provide the additional information required for defining a node which can be launched in a composable context. 

#### Constructor

Most parameters would be passed to the superclass.

|Argument|Description|
|---|---|

> | package        | Optional. Name of the ROS package the node plugin lives in. If not specified, the package of the containing executable will be assumed.         |

Reviewer 2: @roger-strain I like the feature, but when loading a composable node into an already running executable, you may not necessarily have access to package names. So we should probably mention that this deduction may fail.

Author: I'm adding a note to that effect, but as I think about it I worry that we may not have that information more often than not. Might have to determine whether this is feasible at implementation time.

|Argument|Description|
|---|---|
|nodes|Collection of `launch_ros.descriptions.ComposableNode` objects that should be launched inside the `launch_ros.descriptions.Node` with this trait.|

#### Properties

Accessors would be provided for the constructor parameter: `nodes`.

#### Methods

| Name    | Description                                                  |
| ------- | ------------------------------------------------------------ |
| prepare | Overridden. Adds an event handler to launch the defined composable nodes in the container once the container is started. |

#### Events

No events would be handled or emitted.

### launch_ros.descriptions.RosExecutable

This class would represent the information required to run a ROS-aware process and inherit from `launch.descriptions.Executable`.
It would not include execution-time details, nor provide monitoring of the process after launch. 

#### Constructor

| Argument        | Description                                                  |
| --------------- | ------------------------------------------------------------ |
| package        | name of the ROS package the node executable lives in         |

> |node_executable|Name of the executable to find|

Reviewer 2: @roger-strain nit: 
```suggestion
| executable_name |Name of the executable to find|
```
?

Author: Reasonable. Was just copying/pasting parameter names from existing things, I think.

| nodes             | A list of `launch_ros.descriptions.Node` objects which are part of the executable. |

> Additional parameters that may be passed, which are handled by `launch.actions.ExecuteProcess`: `cmd`, `name`, `cwd`, `env`, `additional_env`.

Reviewer 2: @roger-strain consider:
```suggestion
Additional parameters that may be passed, which are handled by `launch.actions.Executable`: `name`, `cwd`, `env`, `additional_env`.
```

While the current refactoring will not attempt to implement this functionality, such functionality could be implemented using sibling classes to those described above.

### launch.actions.ExecuteSshProcess

Extends the `launch.actions.Action` class.
This class would represent the execution-time aspects of an analog to `launch.actions.ExecuteProcess` which executes the given process via an SSH connection.
This likely would be a remote machine in virtually all usages. This is *not* a subclass of the proposed `launch.actions.ExecuteLocal` class; certain features provided by that class are not available from standard SSH connections.

It is also conceivable that this class, or one like it, could be constructed to simultaneously execute multiple processes over a single SSH connection.
If this is attempted, several features such as input/output pipes would likely no longer be feasible.

#### Constructor

| Argument               | Description                                                  |
| ---------------------- | ------------------------------------------------------------ |
| process\_description   | The `launch.descriptions.Executable` to execute as a remote process |

> | connection_description | An object defining the SSH connection information, such as host, port, user, etc. |

Reviewer 2: @roger-strain hmm, one thought (nothing to be changed): if `connection_description` were to become an interface to a given execution environment, a single `launch.actions.Execute()` could later deal with all forms of execution generically. May be overkill though.

Author: @hidmic Wanted to follow up on this, as I was having some thoughts on that topic and wanted to run an idea by you before I start writing it up in the formal document. What would you think about going ahead with `launch.actions.Execute()`, which takes two things: the `launch.descriptions.Executable` to actually run, and *optionally* a `launch.descriptions.ExecutionEnvironment` or subclass thereof, which describes how it should be executed. If you pass nothing, local execution is assumed, and a default `launch.descriptions.LocalExecutionEnvironment` object is created under the hood.

You could inspect the object to get a reference to that default object later, if necessary. If you wanted to set any of the parameters currently assigned to `launch.actions.ExecuteLocal`, you would instead create your own `launch.descriptions.LocalExecutionEnvironment` with the appropriate bits.

(All of the following is forward-thinking and not part of the scope of this refactor):
This then would allow you to also define things like `launch.descriptions.SshExecutionEnvironment`, which would open an SSH session using the various parameters passed to it, or `launch.descriptions.PsExecutionEnvironment` if you have something with a PowerShell flavor. Where this could really shine is if you wanted to do something fancy, like have a single SSH session which could run multiple processes. That could look like a `SharedSshExecutionEnvironment` that you pass to all the relevant `launch.actions.Execute` calls, and then at the end a `SharedSshExecutionEnvironment.ExecuteAll()` method would let you launch one SSH session and run all the things that had been queued up inside it.

I think I'd need to mull a little more about where certain things end up, but it seemed like an interesting approach, and I wanted to capture my thoughts and start the discussion before I forgot.

Reviewer 2: At first sight, all of that makes sense to me. If I may, I'd also suggest a `Group`-like action to easily put multiple actions in the same execution environment (akin to `roslaunch`'s `<machine/>` tag).

Author: @hidmic After poking around at this for a bit, I'm not sure we're really gaining anything with that aspect of the refactor. Basically everything that is defined in the currently proposed `launch.actions.ExecuteLocal` would have to be migrated into the new `launch.descriptions.LocalExecutionEnvironment`, and several methods which are currently overrides of `launch.actions.Action` would have to be redefined and reimplemented on a base `launch.descriptions.ExecutionEnvironment`. The new `launch.actions.Execute` class would basically be a nearly-empty shell that reroutes its own methods into calling the same-named methods on the `launch.descriptions.ExecutionEnvironment` class. It kind of feels like just adding an extra layer with little to no real benefit.

If it turns out that there would be benefit later, such as when trying to add multi-machine support, we could address it at that point.

Curious as to your thoughts. Assuming we're okay just staying with the design as it stands, do we need any further feedback from other contributors, or is it time to start implementation?

# Refactoring ExecuteProcess in ROS2

Initially proposed in [114](https://github.com/ros2/launch/issues/114), this document describes changes to `ExecuteProcess` and related classes.

> As initially implemented, some aspects of this are tightly coupled to both the implementation chosen and the assumption that processes will execute only in the local environment of the launch process itself.

Reviewer 2: @roger-strain nit: 
```suggestion
As initially implemented, some aspects of this class are tightly coupled to both the implementation chosen and the assumption that processes will execute only in the local environment of the launch process itself.
```

> The following changes are designed to provide a more decoupled architecture that will ease support of launching processes locally, remotely, or potentially in containerized environments.

Reviewer 2: @roger-strain nit: 
```suggestion
The following changes are designed to provide a more decoupled architecture that will ease support of launching processes locally, remotely, and potentially in containerized environments.
```

Containerization isn't mutually exclusive with execution in a remote host :)

`ComposableNode` provides a basis for similar breakouts of other elements.

In general, existing classes should have similar analogs in the refactor to ensure that existing classes can be replaced with backwards-compatible replacements that use the new structure.
These replacements can be implemented to warn the user that the older definitions are deprecated, and suggest moving to the newer refactored style.

Further consideration will be required for alternate frontends, such as XML, to determine how to implement these changes with streamlined syntax where possible.

One difficulty when looking forward to allowing functionality such as remote and containerized execution is that the target environment is likely different than the local environment.
Different packages may be installed, different environment variables may be defined, and so forth.
An subclass implementation of the proposed `launch.descriptions.Executable` class may choose to override the base `apply_context` method to implement additional layer(s) of substitutions appropriate to its destination contexts, such as remote installation paths or environment variables.

> Additionally, the available support for concepts such as signals and PIDs may be different; to that end, it is likely that remote execution of processes will require sibling classes to those described below, rather than direct subclasses.

Reviewer 2: @roger-strain hmm, this is a good point. To some extent, I'd expect Python standard library to abstract us away from that. I think we should try to keep OS idiosyncrasies an implementation detail, preferably in an upstream package :)

Author: Agree with you here; mostly it's an explanation of why some things that might normally be assumed to be part of an executable might need to be restricted to a more specific context. Assuming we get to the point of implementing non-local execution (which is where we want to go next) some of this design may have be reconsidered/adjusted as we find out which things really are and are not common. But that's a discussion for the future, I think. If there's anything that immediately seems like it's in the wrong spot, please let me know. :)

#### Properties

Accessors would be provided for the various constructor parameters: `cmd`, `name`, `cwd`, `env`, and `additional_env`.

#### Methods

In the current classes, substitutions are handled through internal method calls.
To support better isolation of class functionality, substitutions will be performed in this class, but exposed from a public method which can be overridden by subclasses to perform additional appropriate substitutions.
Substituted versions of the various parameters will be stored in a local dictionary, similar to the current `__process_event_args`.

| Name          | Description                                                  |
| ------------- | ------------------------------------------------------------ |
| apply_context | Takes a given LaunchContext and performs actions necessary to prepare to execute the defined command in that context. This will primarily consist of expanding various substitutions. |

#### Events

No events would be handled nor emitted.

### launch.actions.ExecuteLocal

Extends the `launch.actions.Action` class.
This class would represent the execution-time aspects of `launch.actions.ExecuteProcess`.

> The new `process_description` constructor parameter could be a `launch.descriptions.Executable`, a `launch_ros.descriptions.Node`, or one of the other subclasses.

Reviewer 2: @roger-strain why do executing actions have to deal with anything but an `Executable` (or derived classes)?

Author: Good catch, removing reference to `Node` here.

| `launch.events.process.ProcessStarted` | Emitted when the process starts                              |
| `launch.events.process.ProcessExited`  | Emitted when the process exits; event contains return code   |
| `launch.events.process.ProcessStdout`  | Emitted when the process produces data the stdout pipe; event contains the data from the pipe |
| `launch.events.process.ProcessStderr`  | Emitted when the process produces data the stderr pipe; event contains the data from the pipe |

### launch_ros.descriptions.Node

This class would represent only the common information required to define a ROS2 node.
It would not inherit from any other class, and would have a subset of the fields of the current `launch_ros.actions.Node`.

#### Constructor

These parameters are drawn from the shared parameters of the current `launch_ros.actions.Node` and `launch_ros.descriptions.ComposableNode`, and introduces a new collection of `traits`, described in following classes.

| Argument       | Description                                                  |
| -------------- | ------------------------------------------------------------ |

> | package        | name of the ROS package the node executable lives in         |

Reviewer 2: @roger-strain hmm, I'd rather say ROS executables and node plugins live in packages, not nodes. So I'm not sure if it makes sense for node descriptions to refer to a package.

Author: I've adjusted this now as well. `package` has moved to the `RosExecutable` object, and been re-introduced for `ComposableNode`. I also want the one on `ComposableNode` to be optional, so it can be picked up from the executable context if possible. If that's not reasonable, we can remove that concept. Would be nice to not have to respecify the package for each composed node. As I'm writing that, it feels wrong, so I think it might have to come back out. Let me know what you think.

Reviewer 2: > I also want the one on ComposableNode to be optional, so it can be picked up from the executable context if possible. If that's not reasonable, we can remove that concept. Would be nice to not have to respecify the package for each composed node.

That's interesting sugar. I like it.

#### Constructor

Most parameters would be passed to the new superclass.

|Argument|Description|
|---|---|
|node_executable|Name of the executable to find|

#### Properties

Accessors would be provided for the constructor parameter: `node_exectuable`.

#### Methods

No custom methods would be defined or overridden.

#### Events

No events would be handled or emitted.

### launch_ros.descriptions.RosExecutable

This class would represent the information required to run a ROS-aware process and inherit from `launch.descriptions.Executable`.
It would not include execution-time details, nor provide monitoring of the process after launch. 

#### Constructor

| Argument        | Description                                                  |
| --------------- | ------------------------------------------------------------ |
| nodes             | A list of `launch_ros.descriptions.Node` objects which are part of the executable. |

Additional parameters that may be passed, which are handled by `launch.actions.ExecuteProcess`: `cmd`, `name`, `cwd`, `env`, `additional_env`.

> NOTE: To allow ROS to determine the appropriate executable based on the package and executable name, the `cmd` parameter should *NOT* be specified. When `cmd` is not specified, this class will determine the appropriate executable if and only if exactly one of the nodes it contains has the `launch_ros.traits.IsExecutable` trait.

Reviewer 2: @roger-strain why can't the `IsExecutable` trait be subsumed into this ` Executable` subclass, taking package name and executable name as arguments (besides node descriptions)? The current `Node` equivalent would then be:

```python
ExecuteLocal(
   ROSExecutable(
       package_name='my_package',
       executable_name='my_node_executable',
       nodes=[
           Node(name='my_node')
       ]
   )
)
```

Author: I like this. I've made that change (though the parameter names may need to be tweaked). See how it looks.

Reviewer 2: I think we can drop the note entirely. `cmd` is not a valid argument for `RosExecutable`'s constructor. No need to ask for it to not be used. If it comes along, `TypeError: invalid arguments` should be raised.
